### ISSUE TITLE
Concatenate non identical sources for ACCNAME and AccDescription Computation 

### ISSUE THREAD

**Issue by DavidMacDonald** [w3c/accname#14]
Here's a thread

## Bryan says:
I'd like to verify the expectation for the following testable statement.

```<input type="text" aria-label="LabelText" title="TitleText" id="test" aria-describedby="descid"><div id="descid">DescriptionText</div>```
Current results:
```accName: "LabelText"````
``accDesc: "DescriptionText"```
The title is lost. Is this the expected result?

## Stephan says
I understood this as override accepted by the group ever since the very beginning. 
Not clear where exactly this is handled in the accessible description computation requirements.
Personally, I feel unhappy with this override.

## David Says
It is the expected result because the title is a swing attribute that fills an empty accname or accdesciption
My preference would be that if it is different then it is a concatenation with the accdescrption...
There a lot if override errors in the wild

## Bryan says
 I understand, and have always thought the invisible loss of accessible information was strange.
Should this be raised as an issue to be added to the naming computation, or should this be ignored as something everybody knows and just deals with?
>“The trick is redundancy... can the user agents recognize identical strings of information and not pass that through”

If this is dealt with in the spec, I’m sure this can be done, since it is already happening in some browsers when the description and name properties match, so description is suppressed so that only name is set. Technically this is possible, it just needs to be dealt with in the algorithm.

## David Says: 
I filed a [bug against Chrome](https://bugs.chromium.org/p/chromium/issues/detail?id=829352&can=2&q=os%3Dchrome) to have them show the ```title``` attribute on tab keyboard focus... 
With respect to the original question there is also the same issue with aria-errormessage and aria-describedby I expect.
The fact is in the real world the title attribute is used a lot... and if browsers can suppress a redundant title (or aria-describedby etc.) and let additional content pass through, this would be desirable... 

# The suggestion

Change the computation to let **non redundant** sources pass through the accessibility tree and be concatenated to the AccName or AccDescription as appropriate.

- if an html label and an aria-labelledby are used, they will be concatenated in the accname if they are not identical.
- If an aria-error-message and aria-describedby are on a field they will both be concatenated in the accdescription
- if a title attribute offers more information and there is already an accdescription then it will be concatenated if it is not identical.

**Comment by joanmarie** [comment]
@DavidMacDonald ok, now I'm confused. You moved this from the ARIA issues (good and thank you). But you closed this one?

**Comment by DavidMacDonald** [comment]
ooops

**Comment by accdc** [comment]
• if an html label and an aria-labelledby are used, they will be concatenated in the accname if they are not identical.
• If an aria-error-message and aria-describedby are on a field they will both be concatenated in the accdescription
• if a title attribute offers more information and there is already an accdescription then it will be concatenated if it is not identical.

I see the need for clarifying what to do with title + aria-describedby and possibly aria-errormessage if no other property is meant for this in the mappings, but I think we will be opening a huge can of worms with aria-label and aria-labelledby because this will change everything about how the computation is calculated. It seems like combining title + aria-describedby is a simple positive change, but the label ones I think are going to need a lot more support by browser implementors to consider because this will change everything that they currently have under the hood.

**Comment by mcking65** [comment]
Title is like a fallback in the calculation: only use if there is nothing better, where there is an assumption that accessibility specific attributes are meant for assistive technologies and elements like title are aimed at mouse users.

In practice, user agents will not be able to avoid creating huge amounts of redundancy and extra verbage for screen readers just with a simple string compare. I think we could create a huge problem of tons of unwanted and super annoying speech if we were to concatenate. It could be dreadful. I do not support automatic concatenation.

That said, I too do not like the idea of lost information. So, if the title attribute is not easily accessible by AT as a separate entity,, then perhaps we should fix that problem so that AT users always have an easy way to get to the title. We have a tooltip role, but perhaps we should have a tooltip property that is mapped to the title string.

**Comment by devarshipant** [comment]
> if an html label and an aria-labelledby are used, they will be concatenated in the accname if they are not identical.

Correct me  -- I think if label and aria-labelledby are present on a control, aria-labelledby wins. 

[Here is a test page.](https://codepen.io/devpant/full/WJNKBK/)

**Comment by DavidMacDonald** [comment]
@devarshipant 
That's right...  I think @accdc was just restating the proposal, not the current status.

**Comment by accdc** [comment]
Hi,
Just to revisit this since a lot of time has passed and we've discussed this at various times in the WG and have reached what seems to be an agreement about how the mappings should treat this for now.

I wanted to map these things out with a very basic overview.

The algorithms are actually the same between aria-labelledby and aria-describedby, both follow the same mechanism for concatenating multiple ID references and apply the same calculations when parsing embedded content.

The only difference here is which property is being written to. E.G The accessible name is processed and saved as the Name property, then the description is processed and saved in the Description property. 

The aria-describedby attribute is always used in reference to setting the Description property, and never applies to setting the Name property.

However, the title attribute applies to both of these properties, with the following exceptions.

# When there is no Name, as previously set with a valid naming mechanism, then the title attribute is set as the Name property when present, and no Description property is set.
# When a valid Name is already set however, then the title attribute is set as the Description property instead.
# When aria-describedby is set, then the referenced computation, if it is not null and does not include only white space characters, is set as the Description property; regardless what the title attribute contains.
# If, after both the accessible Name and Description properties are computed, and if both Name and Description include the same exact value, then both should be merged into the Name alone to prevent redundant duplication by assistive technologies.

This is a very high level summary of what we've discussed. Many of these processes are already covered by the current algorithm, and others will need to be clarified further. This is meant just to illustrate the core relationship between the Name and Description properties, and how they sometimes appear to overlap when applied to the root node (the primary element being processed for its accessible Name and Description).

Does this make sense? Please let me know if I'm missing anything.
